File: 49 Group Anagrams/Solution.py
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        grouped_dict = defaultdict(list)
        for word in strs:
            keyArr = [0] * 26
            for char in word:
                keyArr[ord(char) - 97] += 1
            #cant use ''.join(map(str,keyArr)) because
            # 1b and 10d will lead to 0 1 0 10  and
            # 10b 1c would lead to    0 10 1 0 which is same number
            key = tuple(keyArr)
            grouped_dict[key].append(word)
        return grouped_dict.values()

if __name__ == "__main__":
    solution = Solution()
    strs = ["bdddddddddd","bbbbbbbbbbc"] #["eat","tea","tan","ate","nat","bat"] #[""]#["a"]
    result = solution.groupAnagrams(strs)
    print(result)

File: 49 Group Anagrams/Solution.java
import java.util.*;

class Solution49 {
    public List<List<String>> groupAnagrams(String[] strs) {
        // Using a HashMap where the key is a tuple-like representation of the character counts
        Map<String, List<String>> groupedMap = new HashMap<>();

        for (String word : strs) {
            // Create an array to count character occurrences
            int[] charCount = new int[26];

            for (char c : word.toCharArray()) {
                charCount[c - 'a']++;
            }

            // Convert charCount array into a string as the key
            StringBuilder keyBuilder = new StringBuilder();
            for (int count : charCount) {
                keyBuilder.append(count).append('#'); // Append with a separator to avoid confusion between numbers
            }
            String key = keyBuilder.toString();

            // Add word to the corresponding list in the map
            groupedMap.computeIfAbsent(key, k -> new ArrayList<>()).add(word);
        }

        // Return all the lists of grouped anagrams
        return new ArrayList<>(groupedMap.values());
    }

    public static void main(String[] args) {
        Solution49 solution = new Solution49();
        String[] strs = {"bdddddddddd", "bbbbbbbbbbc"};//{"eat","tea","tan","ate","nat","bat"};//{""};//{"a"};
        List<List<String>> result = solution.groupAnagrams(strs);
        System.out.println(result);
    }
}